File: src/application/directives.ts
// Frontend mirror of the Python `# smj:` directive parser
// (save_my_jupyter/application/snapshot/directives.py). Behavior must stay
// identical; the shared fixtures in fixtures/directives.json are exercised by
// both this module's tests and the Python tests to guard against drift
// (contracts C-DIRECTIVE-01/02, C-CONTENT-08).

export interface DirectiveResult {
  readonly runLabel: string | null;
  readonly tags: readonly string[];
}

// Only full-line comments are directives, so a `#` inside a string literal on a
// code line cannot masquerade as one (we parse statically, without tokenizing).
const COMMENT_MARKERS = ["#", "//"] as const;
const PREFIX = "smj:";

export function parseDirectives(
  cellSources: readonly string[],
): DirectiveResult {
  let runLabel: string | null = null;
  const tags: string[] = [];
  const seen = new Set<string>();
  for (const source of cellSources) {
    for (const line of source.split(/\r?\n/)) {
      const body = directiveBody(line);
      if (body === null) {
        continue;
      }
      const { run, tags: lineTags } = parseBody(body);
      if (run !== null && runLabel === null) {
        runLabel = run;
      }
      for (const tag of lineTags) {
        if (!seen.has(tag)) {
          seen.add(tag);
          tags.push(tag);
        }
      }
    }
  }
  return { runLabel, tags };
}

/** Union tags across sources, whitespace-trimmed and de-duplicated in
 * first-seen order (contract C-CONTENT-08). */
export function mergeTags(
  ...sources: readonly (readonly string[])[]
): string[] {
  const merged: string[] = [];
  const seen = new Set<string>();
  for (const source of sources) {
    for (const tag of source) {
      const trimmed = tag.trim();
      if (trimmed && !seen.has(trimmed)) {
        seen.add(trimmed);
        merged.push(trimmed);
      }
    }
  }
  return merged;
}

function directiveBody(line: string): string | null {
  const stripped = line.replace(/^\s+/, "");
  for (const marker of COMMENT_MARKERS) {
    if (stripped.startsWith(marker)) {
      const comment = stripped.slice(marker.length).replace(/^\s+/, "");
      if (comment.slice(0, PREFIX.length).toLowerCase() === PREFIX) {
        return comment.slice(PREFIX.length).trim();
      }
      return null;
    }
  }
  return null;
}

function parseBody(body: string): { run: string | null; tags: string[] } {
  let run: string | null = null;
  const tags: string[] = [];
  for (const pair of body.split(";")) {
    const equals = pair.indexOf("=");
    if (equals === -1) {
      continue;
    }
    const key = pair.slice(0, equals).trim().toLowerCase();
    const value = pair.slice(equals + 1).trim();
    if (key === "run") {
      if (run === null && value) {
        run = value;
      }
    } else if (key === "tags") {
      for (const tag of value.split(",")) {
        const trimmed = tag.trim();
        if (trimmed) {
          tags.push(trimmed);
        }
      }
    }
  }
  return { run, tags };
}

File: save_my_jupyter/application/snapshot/directives.py
"""Static parser for in-source `# smj:` directives (target CAPTURE, contracts
C-DIRECTIVE-01/02, C-CONTENT-08). Pure: it reads cell source strings and never
executes anything, so it needs no kernel package, import, or IPython hook. The
frontend mirror in ``src/application/directives.ts`` must match this behavior;
the shared fixtures in ``tests/test_directives.py`` guard that contract."""

from __future__ import annotations

from collections.abc import Iterable, Sequence

from save_my_jupyter.domain.capture import DirectiveResult

# Only full-line comments are directives. Trailing inline comments are ignored so
# that `#` inside a string literal on a code line cannot masquerade as a directive
# (we parse statically, without tokenizing the source).
_COMMENT_MARKERS = ("#", "//")
_PREFIX = "smj:"


def parse_directives(cell_sources: Sequence[str]) -> DirectiveResult:
    run_label: str | None = None
    tags: list[str] = []
    seen: set[str] = set()
    for source in cell_sources:
        for line in source.splitlines():
            body = _directive_body(line)
            if body is None:
                continue
            line_run, line_tags = _parse_body(body)
            if line_run is not None and run_label is None:
                run_label = line_run
            for tag in line_tags:
                if tag not in seen:
                    seen.add(tag)
                    tags.append(tag)
    return DirectiveResult(run_label=run_label, tags=tuple(tags))


def merge_tags(*sources: Iterable[str]) -> tuple[str, ...]:
    """Union tags across sources (directives, UI, config defaults), whitespace-
    trimmed and de-duplicated in first-seen order (contract C-CONTENT-08)."""
    merged: list[str] = []
    seen: set[str] = set()
    for source in sources:
        for tag in source:
            trimmed = tag.strip()
            if trimmed and trimmed not in seen:
                seen.add(trimmed)
                merged.append(trimmed)
    return tuple(merged)


def _directive_body(line: str) -> str | None:
    stripped = line.lstrip()
    for marker in _COMMENT_MARKERS:
        if stripped.startswith(marker):
            comment = stripped[len(marker) :].lstrip()
            if comment[: len(_PREFIX)].lower() == _PREFIX:
                return comment[len(_PREFIX) :].strip()
            return None
    return None


def _parse_body(body: str) -> tuple[str | None, list[str]]:
    run: str | None = None
    tags: list[str] = []
    for pair in body.split(";"):
        if "=" not in pair:
            continue
        key, _, value = pair.partition("=")
        normalized_key = key.strip().lower()
        trimmed_value = value.strip()
        if normalized_key == "run":
            if run is None and trimmed_value:
                run = trimmed_value
        elif normalized_key == "tags":
            tags.extend(tag.strip() for tag in trimmed_value.split(",") if tag.strip())
    return run, tags
